packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -2106,283 +2108,111 @@ namespace Tpetra {
-              ++curPos;
-            }
-          } // for each input column index
-        }
-        else { // global 2-D indexing
-          // mfh 21 Jul 2017: This is not thread safe, because
-          // creating an Teuchos::ArrayView or Teuchos::ArrayRCP that
-          // views a Teuchos::ArrayRCP is not.  We could fix that by
-          // making the left-hand side Teuchos::ArrayRCP<GO>&, but
-          // that would not solve the problem below.
-          Teuchos::ArrayView<GO> gblInds = (this->gblInds2D_[lclRow]) ();
-          // Teuchos::ArrayView::operator() takes (offset, size).
-          //
-          // mfh 21 Jul 2017: This is not thread safe, because
-          // creating a subview of a Teuchos::ArrayView is not.
-          auto gblIndsCur = gblInds (0, rowInfo.numEntries);
-          auto gblIndsNew = gblInds (rowInfo.numEntries,
-                                     rowInfo.allocSize - rowInfo.numEntries);
-
-          size_t curPos = 0;
-          for (size_t k_new = 0; k_new < numInputInds; ++k_new) {
-            const GO gblIndToInsert = inputGblColInds[k_new];
-
-            bool isAlreadyInOld = false;
-            for (size_t k_old = 0; k_old < rowInfo.numEntries; ++k_old) {
-              if (gblIndsCur[k_old] == gblIndToInsert) {
-                isAlreadyInOld = true;
-                break;
-              }
-            }
-            if (! isAlreadyInOld) {
-              TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC
-                (curPos >= numNewToInsert, std::logic_error, "curPos = " <<
-                 curPos << " >= numNewToInsert = " << numNewToInsert << ".  "
-                 "Please report this bug to the Tpetra developers.");
-              gblIndsNew[curPos] = gblIndToInsert;
-              ++curPos;
-            }
-          } // for k_new in 0 .. numInputInds - 1
-        } // whether 1-D or 2-D indexing
 
-        this->k_numRowEntries_(lclRow) = rowInfo.numEntries + numNewToInsert;
-        this->setLocallyModified ();
-
-        if (debug) {
-          newNumEntries = rowInfo.numEntries + numNewToInsert;
-          const size_t chkNewNumEntries = this->getNumEntriesInLocalRow (lclRow);
-          TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC
-            (chkNewNumEntries != newNumEntries, std::logic_error,
-             "After inserting new entries, getNumEntriesInLocalRow(" << lclRow <<
-             ") = " << chkNewNumEntries << " != newNumEntries = " << newNumEntries
-             << ".  Please report this bug to the Tpetra developers.");
-        }
-        return numNewToInsert; // all done!
-      } // if the graph is StaticProfile
-      else { // DynamicProfile
+    if (this->getProfileType () == StaticProfile) {
+      auto numEntries = rowInfo.numEntries;
+      using inp_view_type = View<const GO*, execution_space, MemoryUnmanaged>;
+      inp_view_type inputInds(inputGblColInds, numInputInds);
+      auto numInserted = Details::insertCrsIndices(lclRow, k_rowPtrs_,
```

Comment:
It's a bit annoying that this function returns signed, but the caller returns unsigned.  I'm not too bothered by that, though.  `std::optional` and ilk could help here, once they arrive.

Reply:
The signed return value was for the function to return a negative number to indicate 1 or more indices were invalid.  I can change it to be unsigned and return OrdinalTraits<size_t>::invalid() for that case instead

Reply:
Changed in latest commit